onnx/version_converter/convert.h

```diff
@@ -0,0 +1,160 @@
+        OpSetID(7), OpSetID(6)));
+    }
+
+    ModelProto convert_version(
+        const ModelProto& mp_in,
+        const OpSetID& initial_version,
+        const OpSetID& target_version) const override {
+      const std::string initial_domain = initial_version.domain();
+      const std::string target_domain = target_version.domain();
+
+      ONNX_ASSERTM((initial_domain == "" || initial_domain == "ai.onnx") &&
+        (target_domain == "" || target_domain == "ai.onnx"),
+          "Warning: default onnx version converter can only convert "
+          " between default domain opset versions ('' or 'ai.onnx')\n"
+          "Provided initial_domain: %s, provided target_domain: %s",
+          initial_domain.c_str(), target_domain.c_str());
+
+      ONNX_ASSERTM(initial_domain == target_domain,
+        "initial_version and target_version must have the same domains");
+      for (auto it = mp_in.opset_import().begin(); it != mp_in.opset_import()
+          .end(); ++it) {
+        if (it->domain() == initial_version.domain()) {
+          ONNX_ASSERTM(initial_version.version() == it->version(),
+              "initial_version does not reflect current state of model");
+        }
+      }
+
+      std::shared_ptr<Graph> g(ImportModelProto(mp_in));
+      ONNX_ASSERTM(g.get() != nullptr,
+        "Warning: onnx version converter is unable to parse input model. "
+        "(The IR version of the ONNX model may be too old.)");
+
+      // TODO: Move to Inter-Domain Converter
+      // Get initial model versions
+      // std::vector<OpSetID> initial_versions = g->opset_versions_mutable();
+
+      // No conversion necessary if Model has single, equivalent opset version
+      // if (initial_versions.size() == 1 && initial_versions[0].version ==
+      //    target_version.version && initial_versions[0].domain ==
+      //    target_version.domain) {
+      //  return mp_in;
+      // }
+
+      // Check if target_version is valid
+      const std::unordered_map<std::string, std::pair<int, int>>& versions_map = OpSchemaRegistry::DomainToVersionRange::Instance().Map();
+      const std::string search_domain = target_version.domain() == "ai.onnx" ? "" : target_version.domain();
+      const std::pair<int, int>& version_range = versions_map.at(search_domain);
+      ONNX_ASSERTM(target_version.version() >= version_range.first && target_version
+          .version() <= version_range.second,
+        "Warning: invalid target_version (must be between %s and %s",
+        version_range.first, version_range.second);
+      // Compile list of all ops used in the model
+      graph_node_list nodes = g->nodes();
+
+      const std::vector<OpSchema>& all_opschemas = OpSchemaRegistry::get_all_schemas_with_history();
+
+      // Create Map for All Versions of format {op_name: {domain: {version: schema}}}
+      std::unordered_map<std::string, std::unordered_map<std::string, std::map<int64_t, const OpSchema*>>>  all_schemas;
+
```

Comment:
You can iterate the schemas here, and maintain a map [opname==> first_version]. With this, you can automatically create and register `NoPreviousVersionAdapter`.

Reply:
Perfect. This makes your initial point much clearer.